TUF Graduation PR

JustinCappos opened the pull request.

quinton-hoole-2 commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,64 @@
+
+### Document that it is being used successfully in production by at least three independent end users which, in the TOC’s judgement, are of adequate quality and scope.
+
+* TUF is used in production by Cloudflare, Datadog, DigitalOcean, Docker, IBM, Kolide, Microsoft, RedHat, VMware, and many others.  [Uptane](https://uptane.github.io/), the automotive variant of TUF, is sold by major automotive vendors and is being adopted by automakers with over 1/3 of the cars on US roads in their forthcoming cars.  Uptane is included as part of [Automotive Grade Linux](https:/ ...
```
Is it the specification, or the reference implementation thereof that is used in production by these adopters?  I think the answer to that is probably contained somewhere in the links under https://theupdateframework.github.io/adoptions.html, but it would be good to summarize the answer here.

JustinCappos replied:
Added in https://github.com/cncf/toc/pull/166/commits/822588a8bf4a2da22972ba12cdd6ce790df08c40

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,64 @@
+
+As an intentionally minimal security specification / framework, we deliberately do not have a high degree of feature additions in the project.  Effort comes on either the implementations, such as Notary or the Python reference implementation, or the specification itself.  
+
+* Python reference implementation / specification (8 committers, 2 organizations)
```
We've kept the version number as 1.0 (Draft) for basically a year, but it has been settled modulo typo fixes.  I chatted with the other maintainers and we're removing the (Draft) tag.

JustinCappos replied:
Yes, these locations are correct.

quinton-hoole-2 replied:
Great, thanks.

quinton-hoole-2 commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,65 @@
+
+### Document that it is being used successfully in production by at least three independent end users which, in the TOC’s judgement, are of adequate quality and scope.
+
+* TUF is used in production by Cloudflare, Datadog, DigitalOcean, Docker, IBM, Kolide, LEAP, Microsoft, RedHat, VMware, and many others.   [Uptane](https://uptane.github.io/), the automotive variant of TUF, is sold by major automotive vendors and is being adopted by automakers with over 1/3 of the cars on US roads in their forthcoming cars.  Uptane is included as part of [Automotive Grade Linux]( ...
```
Could you please add some additional detail on at least three of these production use cases to assist the TOC in evaluating the "quality and scope" aspect of this requirement. Details like the exact product/service (as opposed to just the company name), what exactly TUF is used for in that product/service, the size of the largest deployments in which TUF is used in those services (in terms of infrastructure size, usage/user volume, transaction volume, etc). And any other info that you think will assist the TOC in gauging "scope and quality".

JustinCappos replied:
The linked ADOPTIONS.md refers to articles and webpages that have what information we have (which is the public blog posts, etc. from the adopting companies talking about their use).  Notary is used in many of the cloud native scenarios.  In the past, when we've asked for data to analyze proposed changes, etc. it is anonymized for privacy reasons and there are multiple rounds of review, sometimes taking months to get data.  (See: https://www.usenix.org/system/files/conference/atc17/atc17-kuppusamy.pdf and https://www.usenix.org/system/files/conference/nsdi16/nsdi16-paper-kuppusamy.pdf, for example.)   

I don't have permission to speak on behalf of any of the adopters and I don't have the level of data you are asking for in this request.  I'm not sure how to proceed.

justincormack replied:
For the Notary production use cases I can fill in some details.

- Docker. We run Notary alongside the Docker public registry with public availability, and have done for some years. All the official library images are signed, as are images we use as part of our products, for example the Docker Enterprise images that we ship, and many other internal and external projects. We also ship Notary as part of the commercial Docker Trusted Registry product which is part of Docker Enterprise.
- IBM. IBM have run Notary for providing content trust on IBM container registry, on their public cloud https://console.bluemix.net/docs/services/Registry/registry_trusted_content.html#registry_trustedcontent
- Microsoft. Azure container registry recently shipped TUF support (I presume this is Notary based) https://docs.microsoft.com/en-us/azure/container-registry/container-registry-content-trust
- Harbor, now a CNCF incubator project, includes Notary support.
- Quay, the CoreOS/Red Hat container registry supports Notary.

lizrice replied:
There is also now an [AWS implementation](https://github.com/awslabs/tough)

iramcdonald commented:
The TUF-variant Uptane is an extremely important technology in the automotive space, providing a secure way for software errors in automobiles to be securely fixed in the field.  TUF provides the back-end security for Uptane and is essential to making the system work in a secure way so that even if an attacker compromises a server or a key, there will not be a loss of life.  With at least one major OEM adopter in the US, Japan, and Europe, in a few years Uptane will be on about a third of new cars on US roads and is quickly on its way to becoming the de facto industry standard.

omgclouds commented:
Looks good to me.

trishankatdatadog commented:
To reiterate @iramcdonald and @abecherer, TUF has been deployed to protect diverse environments ranging from the cloud ([Datadog](https://www.datadoghq.com/blog/engineering/secure-publication-of-datadog-agent-integrations-with-tuf-and-in-toto/), Docker, IBM, Microsoft, Red Hat) to [automotives](https://uptane.github.io/adoptions.html).

I am heavily involved with the project, but this should not discount the conclusion that it is high time for TUF to graduate on the CNCF. Please let us know if anything is blocking it, and what we can do to fix it. Thanks very much.

sjhx commented:
As the tech-lead for IBM Cloud Container registry I'd just like to chime in and confirm that we run Notary in production, as a public service and fully support its graduation.

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,65 @@
+
+### Have a public list of project adopters for at least the primary repo (e.g., ADOPTERS.md or logos on the project website).
+
+Our website has an [adoptions page](https://theupdateframework.github.io/adoptions.html) on it that lists the different projects.  We also have an [ADOPTERS.MD](https://github.com/theupdateframework/tuf/blob/develop/docs/ADOPTERS.md) which contains much the same information.
```
```suggestion
Our website has an [adoptions page](https://theupdateframework.github.io/adoptions.html) on it that lists the different projects.  We also have an [ADOPTERS.MD](https://github.com/theupdateframework/tuf/blob/develop/docs/ADOPTERS.md) which contains much of the same information.  [Uptane](https://uptane.github.io/), the automotive version of TUF, also its own [adoptions page](https://uptane.github.io/adoptions.html).

```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,65 @@
+There are 2 TAPs that are currently under consideration (both of which are likely to be accepted) with authors from 4 organizations:
+* [TAP5](https://github.com/theupdateframework/taps/blob/master/tap5.md) by Trishank Karthik Kuppusamy (Datadog / NYU), Sebastien Awwad (NYU), Evan Cordell (CoreOS / RedHat), Vladimir Diaz, Jake Moshenko (CoreOS / RedHat), Justin Cappos (NYU)
+* [TAP8](https://github.com/theupdateframework/taps/blob/master/tap8.md) by Hannes Mehnert (OCaml / Cambridge University), Justin Cappos (NYU)
+
```
```suggestion

There are also TAPs from the automotive industry dealing with upgrading while changing the TUF metadata version itself and about specifying wireline formats that are currently being iterated over.

```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,73 @@
+
+We have had active contributions from an array of contributors across the CNCF landscape, more traditional software projects, and automakers.  One way to see this is via the substantial changes that made their way into the specification.
+
+Changes to the TUF standard largely come in the form of TAPs (TUF Augmentation Proposals).  There are 7 active, non-informational TAPs (TAP 3-6 and 8-10).  
```
```suggestion
Changes to the TUF standard largely come in the form of TAPs (TUF Augmentation Proposals).  There are 8 active, non-informational TAPs (TAP 3-6 and 8-11).  
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,74 @@
+
+These demonstrate the breadth of contributions from different adopters and the way in which contributions are merged into the specification.
+
+Each part of TUF has seen activity both from core committers and external parties.  Of course the fewest commits have come in the TAPs and spec itself.   The TAPs have had 575 commits by 10 contributors (5 external, 2 were non-TAP authors).  The specification has had 82 commits by 11 contributors (6 external contributors).  Notary has had more than 2.8K commits from 80 contributors (55 external). ...
```
```suggestion
Each part of TUF has seen activity both from core committers and external parties.  Of course the fewest commits have come in the TAPs and spec itself.   The TAPs have had 575 commits by 10 contributors (5 external, 2 were non-TAP authors).  The specification has had 82 commits by 11 contributors (6 external contributors).  Notary has had more than 2.8K commits from 80 contributors (55 external).  TUF's Python reference implementation has had more than 3.3K commits, by 42 contributors (26 external).   This demonstrates both activity and committer diversity.
```

JustinCappos replied:
changes from Santiago...

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,74 @@
+### Document that it is being used successfully in production by at least three independent end users which, in the TOC’s judgement, are of adequate quality and scope.
+
+* TUF is used in production by Cloudflare, Datadog, DigitalOcean, Docker, IBM, Kolide, LEAP, Microsoft, RedHat, VMware, and many others.   [Uptane](https://uptane.github.io/), the automotive variant of TUF, is sold by major automotive vendors and is being adopted by automakers with over 1/3 of the cars on US roads in their forthcoming cars.  Uptane is included as part of [Automotive Grade Linux]( ...
+* While these implementations all use the TUF specification, the implementations vary.  Datadog, OTAinfo, and LEAP use the Python reference implementation, most cloud companies use Notary, Google is integrating their own Go implementation into Fuschia, and most automobiles use custom variants written by different automotive vendors.  
```
```suggestion
* While these integrations all use the TUF specification, the implementations vary.  Datadog, OTAinfo, and LEAP use the Python reference implementation, most cloud companies use Notary, Google is integrating their own Go implementation into Fuschia, and most automobiles use custom variants written by different automotive vendors.  
```

JustinCappos commented:
> There are two main areas that I don't think are fully addressed yet:
> 
> 1. Governance and community
> 
> * https://github.com/theupdateframework/tuf/blob/develop/docs/GOVERNANCE.md - There is currently a "bus factor" of one: if Justin Cappos gets run over by a bus, all governance is out of the window!

Fixed in https://github.com/theupdateframework/tuf/pull/937

> * https://github.com/orgs/theupdateframework/people only has two people listed

We have 19 people, but most are private.  I've asked more folks to switch to public.  You'll see a larger set now and this will increase further as the word gets outt.

> * https://github.com/theupdateframework/specification acknowledgements section and web site at https://theupdateframework.github.io/contact.html both say that the project is managed by Justin Cappos at NYU. The GitHub spec repo makes no mention of CNCF.

Fixed in:
https://github.com/theupdateframework/tuf/pull/939
https://github.com/theupdateframework/theupdateframework.github.io/pull/49
https://github.com/theupdateframework/specification/pull/54

> * Graduated projects are expected to have a vibrant community. There’s a slack channel with 35 members which was entirely silent in 2019 for 8 months, until last month.  The mailing list has 10 threads from this year. Now, perhaps that’s a sign of a stable spec and project, but the community aspect should be documented in the DD. 

We're using slack and mailing lists from resources in my lab which is across many non-CNCF / LF projects as well.  We've added external participants as needed.  @caniszczyk, would the CNCF perhaps be willing to fund paid slack so we can bridge?  (Sorry, we just don't have funds for this now.)  We can also try to move the conversation over, but since many of the members are otherwise non-CNCF participants, it isn't obvious to us what to do.  We're open to changes in how we would do this and how to document it better in the DD.

I'm moving responses to the other half of your questions to the next comment to prevent this from being overly long...

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,74 @@
+* TUF is used in production by Cloudflare, Datadog, DigitalOcean, Docker, IBM, Kolide, LEAP, Microsoft, RedHat, VMware, and many others.   [Uptane](https://uptane.github.io/), the automotive variant of TUF, is sold by major automotive vendors and is being adopted by automakers with over 1/3 of the cars on US roads in their forthcoming cars.  Uptane is included as part of [Automotive Grade Linux]( ...
+* While these integrations all use the TUF specification, the implementations vary.  Datadog, OTAinfo, and LEAP use the Python reference implementation, most cloud companies use Notary, Google is integrating their own Go implementation into Fuschia, and most automobiles use custom variants written by different automotive vendors.  
+* More detail about adoption, including some of the in progress adoptions, can be found here:  https://theupdateframework.github.io/adoptions.html
+
```
```suggestion

Apart from the Docker, LEAP, and AppContainer adoptions (which happened prior to joining the CNCF), these adoptions have all happened while TUF is in incubation.

```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,76 @@
+### Have a public list of project adopters for at least the primary repo (e.g., ADOPTERS.md or logos on the project website).
+
+Our website has an [adoptions page](https://theupdateframework.github.io/adoptions.html) on it that lists the different projects.  We also have an [ADOPTERS.MD](https://github.com/theupdateframework/tuf/blob/develop/docs/ADOPTERS.md) which contains much of the same information.  [Uptane](https://uptane.github.io/), the automotive version of TUF, also its own [adoptions page](https://uptane.github ...
+
```
```suggestion

### Alternatives to TUF

The most common alternative to TUF involves using either a signing key on the server (e.g., TLS) or on a server as part of software creation (e.g., GPG signing in a build farm).  Either way, the fundamental difference is that a single key / server compromise can result in an attacker having the ability to install arbitrary code on end user machines.  Existing specifications / proposals like OMA-DM, SUIT, ITU-T X.1373, as well as common use patterns for GPG/PGP/RSA signing and TLS all have this flaw.  

In contrast, TUF is designed with the assumption that the company deploying TUF may not have perfect security.  In other words, a key and/or server will be compromised at some point.  TUF is focused on preventing an attacker from performing harmful attacks while allowing the operator to recover to a secure baseline.```

JustinCappos commented:
> How has this changed over time while the project has been in incubation?  How has the project matured since incubation?

At the time of CNCF adoption into incubation, the only cloud native production use was in Docker and AppContainer.  We had some footprint outside of the CNCF in LEAP and a few smaller projects.

While in incubation, we have had adoption by Microsoft, Datadog, Google, IBM, RedHat, Cloudflare, DigitalOcean, VMware and many other companies.  The TUF variant Uptane has also been adopted widely outside of the cloud native space, especially in the automotive space where it is an IEEE/ISTO standard and is hosted under the Linux Foundation's Joint Development Foundation.

I've added text to this effect in this PR ( 4ed9c8b8d3d84a8e1b48452249a5821ee3a897d9 ).

> Project and context
> Many of the technical questions don't really apply to a spec project, but we should have some discussion > of context. What are the alternatives? Are there any competing specs in this space?

I've added text to help address this in the PR ( b969c7b6ed27f1bf48fa68f0683b4377d19242ef ).

> Is there a write-up of DD at the incubation stage? That might be helpful.

We didn't do a DD writeup when entering the CNCF because that wasn't part of the process then.  Here is  the PR w/ documentation which may have helpful information.  https://github.com/cncf/toc/pull/38

I hope this helps to clarify things.  Just ask if you need more from us!

lizrice commented:
> We're using slack and mailing lists from resources in my lab which is across many non-CNCF / LF projects as well. We've added external participants as needed. @caniszczyk, would the CNCF perhaps be willing to fund paid slack so we can bridge? (Sorry, we just don't have funds for this now.) We can also try to move the conversation over, but since many of the members are otherwise non-CNCF participants, it isn't obvious to us what to do. We're open to changes in how we would do this and how to document it better in the DD.

I don't particularly think it matters who or where the community is gathering, so long as there is a community, it's open to people to join, and people can find it. Is the mailing list in the NYU lab the same as what's listed on the README? 

I also think it would be OK in the Graduation PR to document the current levels of activity but to acknowledge that since the spec is stable, community activity levels are currently low. But IMO it should be documented so that when TOC members read this PR document to make a judgement on their vote, they know what the situation is.

lizrice commented:
The Governance documentation defines a Consensus Builder and refers to a CB term, but I couldn't find where that term was defined?

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,74 @@
+
+* Python reference implementation / specification (8 committers, 3 organizations)
+
+  * Justin Cappos (NYU), Trishank Karthik Kuppusamy (Datadog / NYU), Lukas Puehringer (NYU), Santiago Torres (NYU), Vladimir Diaz, Sebastien Awwad (Conda), Joey Pabalinas (NYU), Marina Moore (NYU), Lois Anne Delong (NYU)
```
```suggestion
  * Justin Cappos (NYU), Trishank Karthik Kuppusamy (Datadog / NYU), Lukas Puehringer (NYU), Santiago Torres (NYU), Sebastien Awwad (Conda)
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,81 @@
+
+As an intentionally minimal security specification / framework, we deliberately do not have a high degree of feature additions in the project.  Effort comes on either the implementations, such as Notary or the Python reference implementation, or the specification itself.  
+
+* Python reference implementation / specification (8 committers, 3 organizations)
```
```suggestion
* Python reference implementation / specification (5 committers, 3 organizations)
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,81 @@
+
+### Document that it is being used successfully in production by at least three independent end users which, in the TOC’s judgement, are of adequate quality and scope.
+
+* TUF is used in production by Cloudflare, Datadog, DigitalOcean, Docker, IBM, Kolide, LEAP, Microsoft, RedHat, VMware, and many others.   [Uptane](https://uptane.github.io/), the automotive variant of TUF, is sold by major automotive vendors and is being adopted by automakers with over 1/3 of the cars on US roads in their forthcoming cars.  Uptane is included as part of [Automotive Grade Linux]( ...
```
```suggestion
* TUF is used in production by Cloudflare, Datadog, DigitalOcean, Docker, IBM, Kolide, LEAP, Microsoft, RedHat, VMware, and many others.   Amazon has just released a new Rust implementation of [TUF](https://github.com/awslabs/tough).   [Uptane](https://uptane.github.io/), the automotive variant of TUF, is sold by major automotive vendors and is being adopted by automakers with over 1/3 of the cars on US roads in their forthcoming cars.  Uptane is included as part of [Automotive Grade Linux](https://www.automotivelinux.org/), another Linux Foundation project.  
```

lizrice commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,74 @@
+
+* Notary (8 committers, 3 organizations)
+
+  * Ying Li, Diogo Mónica, Evan Cordell (RedHat / CoreOS), David Lawrence, HuKeping (Huawei), Justin Cormack (Docker), Nathan McCauley, Riyaz Faizullabhoy
```
An alternative approach would be to remove the Notary maintainer list from this PR, since you'd like to decouple TUF graduation from Notary's status

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,81 @@
+
+### Have a healthy number of committers. A committer is defined as someone with the commit bit; i.e., someone who can accept contributions to some or all of the project.  The committers should come from at least two organizations.
+
+As an intentionally minimal security specification / framework, we deliberately do not have a high degree of feature additions in the project.  Effort comes on either the implementations, such as Notary or the Python reference implementation, or the specification itself.  
```
```suggestion
As an intentionally minimal security specification / framework, we deliberately do not have a high degree of feature additions in the project.  Effort comes on either the implementations, such as Notary, Amazon's Rust implementation, Google's implementation, or the Python reference implementation, or the specification itself.  As TUF is a specification project, we focus on the specification and the Python reference implementation.
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,81 @@
+
+  * Justin Cappos (NYU), Trishank Karthik Kuppusamy (Datadog / NYU), Lukas Puehringer (NYU), Santiago Torres (NYU), Sebastien Awwad (Conda), Marina Moore (NYU)
+
+* Notary (8 committers, 3 organizations)
```
```suggestion
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,80 @@
+  * Justin Cappos (NYU), Trishank Karthik Kuppusamy (Datadog / NYU), Lukas Puehringer (NYU), Santiago Torres (NYU), Sebastien Awwad (Conda), Marina Moore (NYU)
+
+
+  * Ying Li, Diogo Mónica, Evan Cordell (RedHat / CoreOS), David Lawrence, HuKeping (Huawei), Justin Cormack (Docker), Nathan McCauley, Riyaz Faizullabhoy
```
```suggestion
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,79 @@
+  * Justin Cappos (NYU), Trishank Karthik Kuppusamy (Datadog / NYU), Lukas Puehringer (NYU), Santiago Torres (NYU), Sebastien Awwad (Conda), Marina Moore (NYU)
+
+
+
```
```suggestion
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,78 @@
+
+These demonstrate the breadth of contributions from different adopters and the way in which contributions are merged into the specification.
+
+Each part of TUF has seen activity both from core committers and external parties.  Of course the fewest commits have come in the TAPs and spec itself.   The TAPs have had 575 commits by 10 contributors (5 external, 2 were non-TAP authors).  The specification has had 82 commits by 11 contributors (6 external contributors).  Notary has had more than 2.8K commits from 80 contributors (55 external). ...
```
```suggestion
Each part of TUF has seen activity both from core committers and external parties.  Of course the fewest commits have come in the TAPs and spec itself.   The TAPs have had 575 commits by 10 contributors (5 external, 2 were non-TAP authors).  The specification has had 82 commits by 11 contributors (6 external contributors).  TUF's Python reference implementation has had more than 3.3K commits, by 42 contributors (26 external).   This demonstrates both activity and committer diversity.
```

JustinCappos commented on reviews/tuf-graduation.md:
```diff
@@ -0,0 +1,78 @@
+### Have a public list of project adopters for at least the primary repo (e.g., ADOPTERS.md or logos on the project website).
+
+Our website has an [adoptions page](https://theupdateframework.github.io/adoptions.html) on it that lists the different projects.  We also have an [ADOPTERS.MD](https://github.com/theupdateframework/tuf/blob/develop/docs/ADOPTERS.md) which contains much of the same information.  [Uptane](https://uptane.github.io/), the automotive version of TUF, also its own [adoptions page](https://uptane.github ...
+
```
```suggestion

## Security Audits
There are multiple [security audits](https://theupdateframework.github.io/audits.html) of TUF available on the TUF website.

```

lizrice commented:
Thanks @JustinCappos and TUF team for the revisions on this proposal. I believe TUF meets the graduation requirements so I’d like to call for a vote @caniszczyk @amye

caniszczyk commented:
+1 binding TOC votes (6/9):
Liz Rice: https://lists.cncf.io/g/cncf-toc/message/3777
Jeff Brewer: https://lists.cncf.io/g/cncf-toc/message/3780
Michelle Noorali: https://lists.cncf.io/g/cncf-toc/message/3830
Xiang Li: https://lists.cncf.io/g/cncf-toc/message/3831
Brendan Burns: https://lists.cncf.io/g/cncf-toc/message/3834
Joe Beda: https://lists.cncf.io/g/cncf-toc/message/3872
